investigate: segfault when aborting pytest early with -W error::DeprecationWarning

## Observed behavior

When running the test suite with `-W error::DeprecationWarning` and `-x` (stop on first failure), pytest exits with a segmentation fault after the first failed test:

```
FAILED tests/test_memory.py::TestMemorySearch::test_search_fts_only_skips_vector_path
zsh: segmentation fault  python -m pytest tests/ -q --tb=no -x -W error::DeprecationWarning
```

The full suite (without `-W error`) completes with 1065 passed and no segfault.

## Hypothesis

The segfault likely occurs during pytest teardown when the process aborts early. Possible root causes:

1. A `VstashStore` fixture is not properly closed before the SQLite connection or snapvec/sqlite-vec C extension is garbage-collected — the C extension destructor runs on the wrong thread or against an already-freed handle.
2. The per-thread FTS5 in-memory connection (see `_stem_terms`) is closed from the main thread after the owning thread has exited, hitting a race in the C layer.
3. A `__del__` finalizer on `SnapIndex` or the sqlite-vec virtual table races with pytest's abrupt teardown.

## Steps to reproduce

```bash
python -m pytest tests/ -q --tb=no -x -W error::DeprecationWarning
```

## Investigation checklist

- [ ] Confirm segfault reproduces with `--no-header -p no:randomly` (rule out plugin interference)
- [ ] Run with `PYTHONFAULTHANDLER=1` or `python -X faulthandler` to get a C-level traceback
- [ ] Check whether disabling the snapvec backend (`--no-snapvec` or env var) eliminates the segfault
- [ ] Check whether the segfault disappears when only `test_memory.py` is collected (no other fixtures involved)
- [ ] Review `VstashStore.close()` and `__exit__` to ensure all C-level handles are released before Python GC runs

--- a/tests/test_embed.py
+++ b/tests/test_embed.py
@@ -30,32 +30,34 @@ class TestGetEmbeddingDim:
         assert len(KNOWN_DIMS) >= 4
 
 
-class TestHFOnnxUnavailableLock:
-    """Regression coverage for the threading lock on the shared
-    ``_hf_onnx_unavailable`` set.
+class TestHFOnnxUnavailableConcurrency:
+    """Pin the chosen concurrency semantics of ``_hf_onnx_unavailable``.
 
-    Protects against multiple threads each retriggering a slow ONNX
-    init for a model we already know is broken.  The lock makes the
-    check-then-add pattern observably single-writer; this test just
-    pins that invariant so a future refactor does not silently drop
-    the lock and leave a race behind.
+    The set is deliberately unlocked (set ops are atomic under the
+    GIL; a real guard would have to span the slow ``_init_hf_onnx``
+    call and serialize every fallback).  This test exercises the
+    concurrent fallback path and asserts the two contracts we care
+    about: no thread deadlocks, and the failure marker sticks.
+
+    It does **not** assert "``_init_hf_onnx`` was called exactly
+    once" -- accepting up to ``n_threads`` init attempts is the
+    explicit trade-off for keeping the hot path lock-free.
     """
 
-    def test_concurrent_init_fallback_is_serialised(self, monkeypatch: pytest.MonkeyPatch) -> None:
-        """Spawn N threads that all hit the fallback path for the same
-        model_name.  The set must end with exactly one entry -- the
-        lock serialises the ``.add`` even though each thread sees the
-        check-before-add race individually.
-        """
+    def test_concurrent_init_fallback_is_safe(self, monkeypatch: pytest.MonkeyPatch) -> None:
         import threading
 
         import vstash.embed as embed_mod
 
         # Start clean so we don't leak state into other tests.
-        with embed_mod._hf_onnx_unavailable_lock:
-            embed_mod._hf_onnx_unavailable.discard("broken/model")
+        embed_mod._hf_onnx_unavailable.discard("broken/model")
+        init_calls = 0
+        init_calls_lock = threading.Lock()
 
         def always_fail(model_name: str) -> tuple:
+            nonlocal init_calls
+            with init_calls_lock:
+                init_calls += 1
             raise RuntimeError("simulated ONNX init failure")
 
         monkeypatch.setattr(embed_mod, "_init_hf_onnx", always_fail)
@@ -81,8 +83,17 @@ class TestHFOnnxUnavailableLock:
             t.join(timeout=5)
 
         try:
+            # (1) No thread is stuck -- a future refactor that
+            # introduces a lock around the slow path would show up
+            # here as a join timeout.
+            for t in threads:
+                assert not t.is_alive(), "worker thread did not finish within timeout"
+            # (2) The failure marker is set after the storm.
             assert "broken/model" in embed_mod._hf_onnx_unavailable
-            assert sum(1 for m in embed_mod._hf_onnx_unavailable if m == "broken/model") == 1
+            # (3) Each thread independently tried to init at least
+            # once; we accept up to n_threads attempts (the
+            # deliberately-benign race).  Fail only if zero attempts
+            # happened, which would mean the fallback path never ran.
+            assert 1 <= init_calls <= n_threads
         finally:
-            with embed_mod._hf_onnx_unavailable_lock:
-                embed_mod._hf_onnx_unavailable.discard("broken/model")
+            embed_mod._hf_onnx_unavailable.discard("broken/model")

--- a/tests/test_embed.py
+++ b/tests/test_embed.py
@@ -28,3 +28,61 @@ class TestGetEmbeddingDim:
 
     def test_known_dims_dict_is_populated(self) -> None:
         assert len(KNOWN_DIMS) >= 4
+
+
+class TestHFOnnxUnavailableLock:
+    """Regression coverage for the threading lock on the shared
+    ``_hf_onnx_unavailable`` set.
+
+    Protects against multiple threads each retriggering a slow ONNX
+    init for a model we already know is broken.  The lock makes the
+    check-then-add pattern observably single-writer; this test just
+    pins that invariant so a future refactor does not silently drop
+    the lock and leave a race behind.
+    """
+
+    def test_concurrent_init_fallback_is_serialised(self, monkeypatch: pytest.MonkeyPatch) -> None:
+        """Spawn N threads that all hit the fallback path for the same
+        model_name.  The set must end with exactly one entry -- the
+        lock serialises the ``.add`` even though each thread sees the
+        check-before-add race individually.
+        """
+        import threading
+
+        import vstash.embed as embed_mod
+
+        # Start clean so we don't leak state into other tests.
+        with embed_mod._hf_onnx_unavailable_lock:
+            embed_mod._hf_onnx_unavailable.discard("broken/model")
+
+        def always_fail(model_name: str) -> tuple:
+            raise RuntimeError("simulated ONNX init failure")
+
+        monkeypatch.setattr(embed_mod, "_init_hf_onnx", always_fail)
+        # Make the fallback path a no-op so we don't download ST weights.
+        monkeypatch.setattr(
+            embed_mod,
+            "_embed_hf_st",
+            lambda texts, model_name: [[0.0] for _ in texts],
+        )
+
+        n_threads = 16
+        start = threading.Event()
+
+        def worker():
+            start.wait()
+            embed_mod._embed_hf_onnx(["hello"], "broken/model")
+
+        threads = [threading.Thread(target=worker) for _ in range(n_threads)]
+        for t in threads:
+            t.start()
+        start.set()
+        for t in threads:
+            t.join(timeout=5)
+
+        try:
+            assert "broken/model" in embed_mod._hf_onnx_unavailable
+            assert sum(1 for m in embed_mod._hf_onnx_unavailable if m == "broken/model") == 1
+        finally:
+            with embed_mod._hf_onnx_unavailable_lock:
+                embed_mod._hf_onnx_unavailable.discard("broken/model")

--- a/tests/test_memory.py
+++ b/tests/test_memory.py
@@ -304,10 +304,14 @@ class TestMemorySearch:
 
             # fts_only MUST find the kubernetes-bearing doc regardless
             # of vector distance, because the vector path is skipped.
-            # Assert on the chunk text, not the auto-generated title —
+            # Assert on the chunk text, not the auto-generated title --
             # vstash derives titles from filenames when no frontmatter
             # is present, which is not what this test is about.
-            results = mem.search("kubernetes", fts_only=True)
+            # ``fts_only=True`` is deprecated; wrap in pytest.warns so
+            # the suite stays green under -W error::DeprecationWarning
+            # (#279).
+            with pytest.warns(DeprecationWarning, match="fts_only=True is deprecated"):
+                results = mem.search("kubernetes", fts_only=True)
             assert len(results) > 0
             assert any("kubernetes" in r.text.lower() for r in results), (
                 f"fts_only=True failed to surface the FTS-matching doc; "
@@ -386,8 +390,9 @@ class TestMemorySearch:
 
             # FTS-only on the same query: last_best_distance must
             # be exactly 2.0 (worst case), proving the vector path
-            # was bypassed.
-            mem.search("rust", top_k=3, fts_only=True)
+            # was bypassed.  ``fts_only=True`` is deprecated (#279).
+            with pytest.warns(DeprecationWarning, match="fts_only=True is deprecated"):
+                mem.search("rust", top_k=3, fts_only=True)
             fts_distance = mem._store.last_best_distance
             assert fts_distance == 2.0, (
                 f"fts_only=True should leave last_best_distance == 2.0, got {fts_distance}"
@@ -406,13 +411,15 @@ class TestMemorySearch:
         with Memory(db=tmp_path / "test.db") as mem:
             mem.add(tmp_path / "doc.md")
             # Would raise RRFWeightOutOfRangeError without the
-            # Memory.search override.
-            results = mem.search(
-                "text",
-                fts_only=True,
-                vec_weight=1.5,  # out of range — would normally raise
-                fts_weight=-0.2,  # ditto
-            )
+            # Memory.search override.  ``fts_only=True`` is deprecated
+            # so we expect the corresponding warning too (#279).
+            with pytest.warns(DeprecationWarning, match="fts_only=True is deprecated"):
+                results = mem.search(
+                    "text",
+                    fts_only=True,
+                    vec_weight=1.5,  # out of range -- would normally raise
+                    fts_weight=-0.2,  # ditto
+                )
             assert isinstance(results, list)
 
     @requires_sqlite_vec

--- a/vstash/config.py
+++ b/vstash/config.py
@@ -151,10 +151,11 @@ class StorageConfig(BaseModel):
         ge=0,
         le=1024,
         description=(
-            "IVF coarse clusters for snapvec-ivfpq. 0 = auto (4 * sqrt(N)). "
-            "Explicit values are clamped to [8, 1024] to match the internal "
-            "_nlist_for clamp. FAISS rule: need at least 30 training vectors "
-            "per cluster."
+            "IVF coarse clusters for snapvec-ivfpq. 0 = auto (4 * sqrt(N), "
+            "clamped by ``_nlist_for`` to [8, 1024]). Explicit values must "
+            "fall in [8, 1024]; values outside this range are rejected at "
+            "load time by the field validator. FAISS rule of thumb: need "
+            "at least 30 training vectors per cluster."
         ),
     )
     ivfpq_M: int = Field(

--- a/vstash/config.py
+++ b/vstash/config.py
@@ -153,9 +153,10 @@ class StorageConfig(BaseModel):
         description=(
             "IVF coarse clusters for snapvec-ivfpq. 0 = auto (4 * sqrt(N), "
             "clamped by ``_nlist_for`` to [8, 1024]). Explicit values must "
-            "fall in [8, 1024]; values outside this range are rejected at "
-            "load time by the field validator. FAISS rule of thumb: need "
-            "at least 30 training vectors per cluster."
+            "fall in [8, 1024]: values below 8 are rejected at load time by "
+            "``_check_ivfpq_nlist`` and values above 1024 by Pydantic's "
+            "``le=1024`` field constraint. FAISS rule of thumb: need at "
+            "least 30 training vectors per cluster."
         ),
     )
     ivfpq_M: int = Field(

--- a/vstash/embed.py
+++ b/vstash/embed.py
@@ -493,15 +493,16 @@ _hf_onnx_cache: dict[str, tuple] = {}  # model_name -> (session, tokenizer, max_
 _hf_onnx_lock = threading.Lock()
 _hf_st_cache: dict[str, object] = {}  # model_name -> SentenceTransformer
 _hf_st_lock = threading.Lock()
-# Models whose ONNX init has failed in this process.  Guarded by a
-# dedicated lock so a concurrent read (hot path) and write (cold
-# fallback on first init failure) cannot observe an intermediate
-# set state -- the race is benign under the GIL (a double-failed
-# init is at worst one extra slow-path attempt), but the lock keeps
-# the failure semantics predictable and lines up with the other
-# module-level caches above.
+# Models whose ONNX init has failed in this process.  Deliberately
+# unlocked: ``set.__contains__`` and ``set.add`` are atomic under the
+# GIL, and a real guard against duplicate slow-path init attempts
+# would have to span the ``_init_hf_onnx`` call itself -- which would
+# serialize every fallback embed through a single mutex.  The check-
+# then-add race across threads is observably benign (worst case: two
+# threads both fail to init the same model before either writes the
+# marker; each independently falls back to sentence-transformers for
+# that call, then subsequent calls see the marker).
 _hf_onnx_unavailable: set[str] = set()
-_hf_onnx_unavailable_lock = threading.Lock()
 
 
 def _is_hf_onnx_model(model_name: str) -> bool:
@@ -593,9 +594,8 @@ def _embed_hf_onnx(texts: list[str], model_name: str) -> list[list[float]]:
     if not texts:
         return []
 
-    with _hf_onnx_unavailable_lock:
-        if model_name in _hf_onnx_unavailable:
-            return _embed_hf_st(texts, model_name)
+    if model_name in _hf_onnx_unavailable:
+        return _embed_hf_st(texts, model_name)
 
     try:
         session, tokenizer, max_len = _init_hf_onnx(model_name)
@@ -605,8 +605,7 @@ def _embed_hf_onnx(texts: list[str], model_name: str) -> list[list[float]]:
             model_name,
             exc.__class__.__name__,
         )
-        with _hf_onnx_unavailable_lock:
-            _hf_onnx_unavailable.add(model_name)
+        _hf_onnx_unavailable.add(model_name)
         return _embed_hf_st(texts, model_name)
     all_embeddings: list[list[float]] = []
     batch_size = 32

--- a/vstash/embed.py
+++ b/vstash/embed.py
@@ -493,7 +493,15 @@ _hf_onnx_cache: dict[str, tuple] = {}  # model_name -> (session, tokenizer, max_
 _hf_onnx_lock = threading.Lock()
 _hf_st_cache: dict[str, object] = {}  # model_name -> SentenceTransformer
 _hf_st_lock = threading.Lock()
+# Models whose ONNX init has failed in this process.  Guarded by a
+# dedicated lock so a concurrent read (hot path) and write (cold
+# fallback on first init failure) cannot observe an intermediate
+# set state -- the race is benign under the GIL (a double-failed
+# init is at worst one extra slow-path attempt), but the lock keeps
+# the failure semantics predictable and lines up with the other
+# module-level caches above.
 _hf_onnx_unavailable: set[str] = set()
+_hf_onnx_unavailable_lock = threading.Lock()
 
 
 def _is_hf_onnx_model(model_name: str) -> bool:
@@ -585,8 +593,9 @@ def _embed_hf_onnx(texts: list[str], model_name: str) -> list[list[float]]:
     if not texts:
         return []
 
-    if model_name in _hf_onnx_unavailable:
-        return _embed_hf_st(texts, model_name)
+    with _hf_onnx_unavailable_lock:
+        if model_name in _hf_onnx_unavailable:
+            return _embed_hf_st(texts, model_name)
 
     try:
         session, tokenizer, max_len = _init_hf_onnx(model_name)
@@ -596,7 +605,8 @@ def _embed_hf_onnx(texts: list[str], model_name: str) -> list[list[float]]:
             model_name,
             exc.__class__.__name__,
         )
-        _hf_onnx_unavailable.add(model_name)
+        with _hf_onnx_unavailable_lock:
+            _hf_onnx_unavailable.add(model_name)
         return _embed_hf_st(texts, model_name)
     all_embeddings: list[list[float]] = []
     batch_size = 32

